### ISSUE TITLE
Complex type use not working ?

### ISSUE THREAD

**Issue by eLvErDe** [tefra/xsdata#376]
Hello,

I'm struggling with complexe type, I'm not sure it's a bug, I might just not understand how to use it.
Here is a complete standalone example:

```python
"""
Datex2 complex type not working

Please generate datex2 models first by running:

  xsdata generate https://docs.datex2.eu/_static/data/v3.1/DATEXII_3_Situation.xsd \
    --output pydata --package datexii.datexii_3_1

https://docs.datex2.eu/general/DATEXII-UserGuide-UseOfALERT-C.html

Expected output:

  <alertCLinear xsi:type="AlertCMethod4Linear">
      <alertCLocationCountryCode>[F]</alertCLocationCountryCode>
      <alertCLocationTableNumber>[32]</alertCLocationTableNumber>
      <alertCLocationTableVersion>[alertC table version]</alertCLocationTableVersion>
      <alertCDirection> (i.e. Event direction)
          <alertCDirectionCoded>both (positive, negative)</alertCDirectionCoded>
      </alertCDirection>
      <alertCMethod4PrimaryPointLocation>
          <alertCLocation>
              <alertCLocationName>
                  <value lang="fr">[Name of primary point]</value>
              </alertCLocationName>
          <specificLocation>[Primary point identifier]</specificLocation>
          </alertCLocation>
          <offsetDistance>
              <offsetDistance>[Offset of primary point]</offsetDistance>
          </offsetDistance>
      </alertCMethod4PrimaryPointLocation>
      <alertCMethod4SecondaryPointLocation>
          <alertCLocation>
              <alertCLocationName>
                      <value lang="fr">[Name of secondary point]</value>
              </alertCLocationName>
          <specificLocation>[Secondary point identifier]</specificLocation>
          </alertCLocation>
          <offsetDistance>
              <offsetDistance>[Offset of secondary point]</offsetDistance>
          </offsetDistance>
      </alertCMethod4SecondaryPointLocation>
  </alertCLinear>
"""


import xml.dom.minidom

from xsdata.formats.dataclass.serializers import XmlSerializer

from datexii.datexii_3_1.datexii_3_common import TypeExtensionType
from datexii.datexii_3_1.datexii_3_location_referencing import (
    AlertClinear,
    AlertCdirection,
    AlertCdirectionEnum,
    TypeAlertCdirectionEnum,
    AlertCmethod4Linear,
    AlertCmethod4PrimaryPointLocation,
    AlertCmethod4SecondaryPointLocation,
    AlertClocation,
    OffsetDistance,
)


def build_datex2_3_1_alert_c_linear() -> AlertClinear:
    """
    Build Datex2 AlertC linear position example

    :returns: AlertClinear xsdata object instance
    :rtype: AlertClinear
    """

    loc = AlertClinear(
        alert_clocation_country_code="F",
        alert_clocation_table_number=32,
        alert_clocation_table_version="1.0",
        alert_clinear_extension=TypeExtensionType(
            [
                AlertCdirection(alert_cdirection_coded=TypeAlertCdirectionEnum(value=AlertCdirectionEnum.NEGATIVE)),
                AlertCmethod4PrimaryPointLocation(alert_clocation=AlertClocation(specific_location=10), offset_distance=OffsetDistance(offset_distance=2000)),
                AlertCmethod4SecondaryPointLocation(alert_clocation=AlertClocation(specific_location=12), offset_distance=OffsetDistance(offset_distance=150)),
            ]
        ),
    )
    return loc


if __name__ == "__main__":
    SERIALIZER = XmlSerializer()
    LOC = build_datex2_3_1_alert_c_linear()

    print(xml.dom.minidom.parseString(SERIALIZER.render(LOC)).toprettyxml())
```

Also, I sent you a direct email using the contact form from your personal website.

Regards, Adam.

**Comment by tefra** [comment]
Hi @eLvErDe 

First of all thanks for reporting and providing some sample code  as well, your case it's definitely interesting and I am going to include it in the samples repo to make sure future changes don't break stuff.

Since xsdata 20.12 the generator is creating duplicate class names, for your schema and for schemas that use underscores to define abstract types. I am amazed I haven't seen this pattern before.

```
<complexType name="one">...</complexType>
<complexType name="_one" abstract="true">...</complexType>
```
This results into two classes both named **One**   and I am on it.

Let's take it from the top in the sample output

```<alertCLinear xsi:type="AlertCMethod4Linear">```

The element is named  `alertCLinear` and the type is `AlertCMethod4Linear` which extends `AlertCLinear`. All the names are case sensitive. In the schema there is no root element or complexType named `alertCLinear`.

I am guessing the samples in the docs are describing sub elements/models and they are omitting the content wrapper and all namespaces for readability right?

In the schema there is only one type `SingleRoadLinearLocation` with an attribute `alertCLinear`

```
obj = SingleRoadLinearLocation(
    alert_clinear=AlertCmethod4Linear(
        alert_clocation_country_code="F",
        alert_clocation_table_number=32,
        alert_clocation_table_version="1.0",
        alert_cdirection=AlertCdirection(
            alert_cdirection_coded=TypeAlertCdirectionEnum(
                value=AlertCdirectionEnum.NEGATIVE)
        ),
        alert_cmethod4_primary_point_location=AlertCmethod4PrimaryPointLocation(
            alert_clocation=AlertClocation(specific_location=10),
            offset_distance=OffsetDistance(offset_distance=2000)
        ),
        alert_cmethod4_secondary_point_location=AlertCmethod4SecondaryPointLocation(
            alert_clocation=AlertClocation(specific_location=12),
            offset_distance=OffsetDistance(offset_distance=150)
        )
    )
)

serializer = XmlSerializer(config=SerializerConfig(pretty_print=True))
serializer.write(sys.stdout, obj, ns_map={"loc": "http://datex2.eu/schema/3/locationReferencing"})
```

Which renders like 
```
<?xml version="1.0" encoding="UTF-8"?>
<SingleRoadLinearLocation xmlns:loc="http://datex2.eu/schema/3/locationReferencing">
  <loc:alertCLinear xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="loc:AlertCMethod4Linear">
    <loc:alertCLocationCountryCode>F</loc:alertCLocationCountryCode>
    <loc:alertCLocationTableNumber>32</loc:alertCLocationTableNumber>
    <loc:alertCLocationTableVersion>1.0</loc:alertCLocationTableVersion>
    <loc:alertCDirection>
      <loc:alertCDirectionCoded>negative</loc:alertCDirectionCoded>
    </loc:alertCDirection>
    <loc:alertCMethod4PrimaryPointLocation>
      <loc:alertCLocation>
        <loc:specificLocation>10</loc:specificLocation>
      </loc:alertCLocation>
      <loc:offsetDistance>
        <loc:offsetDistance>2000</loc:offsetDistance>
      </loc:offsetDistance>
    </loc:alertCMethod4PrimaryPointLocation>
    <loc:alertCMethod4SecondaryPointLocation>
      <loc:alertCLocation>
        <loc:specificLocation>12</loc:specificLocation>
      </loc:alertCLocation>
      <loc:offsetDistance>
        <loc:offsetDistance>150</loc:offsetDistance>
      </loc:offsetDistance>
    </loc:alertCMethod4SecondaryPointLocation>
  </loc:alertCLinear>
</SingleRoadLinearLocation>
```

I am working with 21.1 with a fix for the names, so some class names might not match with your generated models but you get the idea.

**Comment by tefra** [comment]
Oh the contact form on my site is broken for at least a few years, please email me directly chris{at}komposta.net

**Comment by eLvErDe** [comment]
Hello @tefra,

Thanks a lot for the quick response. I'm not an XML expert at all, but thanks to the EU commission, I should be very soon !
I can confirm modifying the code to use AlertCmethod4Linear works as expect but sadly the whole model is built on similar patterns so it is quite hard to follow at the moment, so that's a very good new if you can do something.

Thanks, Adam

**Comment by eLvErDe** [comment]
Hello again,

Could you confirm such issue happen with "location_reference" attribute on "SituationRecord" class ?
I expect to see a Union or List[Union but generated models give the same TypeExtensionType...

Thanks !

**Comment by tefra** [comment]
Yeah it's the same situation, think of `LocationReference` as base class, any of the subclasses no matter the level, can be used in these attributes.

![Screenshot from 2021-01-10 10-31-42](https://user-images.githubusercontent.com/2098915/104118168-afa00400-532f-11eb-9b55-0e93574d0a13.png)

The parser understands which exactly type is used through the `xsi:type` attribute and the serializer will make sure to automatically added it. When you are constructing the objects manually you should also check the `xsi:type` to match a class by the fully qualified name.

**Comment by tefra** [comment]
The patch for the names is on master and everything looks in order to me.

The schemas are communication models right, do you have any full samples or a wsdl file, I couldn't find much, all the samples I found are for v2.2 not v3.1

**Comment by eLvErDe** [comment]
Sadly it seems I'm the only one trying to use 3.1 schema (but I'll also provide 2.3 routes) so I'm having the same issue as you...

**Comment by eLvErDe** [comment]
Are you sure you have pushed master ? I see no new patch except unrelated ones

**Comment by tefra** [comment]
Except the duplicate class names I haven't found any other issue @ 95970a6b640b040530fe0726ae7683b25c15daa3

I also suggest using the new naming scheme `mixedPascalCase` for class names so they match a bit better with your schema

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Config xmlns="http://pypi.org/project/xsdata" version="21.1">
  <Output wsdl="false" maxLineLength="79">
    <Package>datexii.models</Package>
    <Format>pydata</Format>
    <Structure>filenames</Structure>
    <DocstringStyle>reStructuredText</DocstringStyle>
    <CompoundFields>false</CompoundFields>
  </Output>
  <Conventions>
    <ClassName case="mixedPascalCase" safePrefix="type"/>
    <FieldName case="snakeCase" safePrefix="value"/>
    <ModuleName case="snakeCase" safePrefix="mod"/>
    <PackageName case="snakeCase" safePrefix="pkg"/>
  </Conventions>
  <Aliases>
    
  </Aliases>
</Config>

```


```python
        obj = SingleRoadLinearLocation(
            alert_clinear=AlertCMethod4Linear(
                alert_clocation_country_code="F",
                alert_clocation_table_number=32,
                alert_clocation_table_version="1.0",
                alert_cdirection=AlertCDirection(
                    alert_cdirection_coded=AlertCDirectionEnum2(
                        value=AlertCDirectionEnum1.NEGATIVE
                    ),
                    alert_caffected_direction=LinearDirectionEnum2(
                        value=LinearDirectionEnum1.OPPOSITE
                    )
                ),
                alert_cmethod4_primary_point_location=AlertCMethod4PrimaryPointLocation(
                    alert_clocation=AlertCLocation(specific_location=10),
                    offset_distance=OffsetDistance(offset_distance=2000),
                ),
                alert_cmethod4_secondary_point_location=AlertCMethod4SecondaryPointLocation(
                    alert_clocation=AlertCLocation(specific_location=12),
                    offset_distance=OffsetDistance(offset_distance=150),
                ),
            )
        )
```

**Comment by eLvErDe** [comment]
Indeed I just checked the generated models:

```python
@dataclass
class SituationRecord:

    location_reference: Optional[LocationReference] = field(
        default=None,
        metadata={
            "name": "locationReference",
            "type": "Element",
            "namespace": "http://datex2.eu/schema/3/situation",
            "required": True,
        }
    )
```

Looks great, thanks a lot

**Comment by eLvErDe** [comment]
Hello @tefra,

A last question regarding this, I'm not sure how to use ns_map, but if you look at the original examples, there's no ns attribute at all. Can you tell me how to set ns_map to render to get such result ?

Thanks in advance,

**Comment by tefra** [comment]
The samples in the site are omitting the namespaces for readability but they are necessary.
